### PR TITLE
dm-master: embed etcd in DM-master

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

embed etcd in DM-master.

### What is changed and how it works?

- add some config items in DM-master which needed by etcd
    - `name`: human-readable name for this DM-master member (default 'dm-master-${hostname}')
    - `data-dir`: path to the data directory (default 'default.${name}')
    - `initial-cluster`: initial cluster configuration for bootstrapping, e,g. 'dm-master=http://127.0.0.1:8291'
    - `peer-urls`: URLs for peer traffic (default 'http://127.0.0.1:8291')
    - `advertise-peer-urls`: advertise URLs for peer traffic (default '${peer-urls}')
- register gRPC API server, HTTP API server and status&debug HTTP server into embed etcd server.

NOTE: we use `master-addr` to generate `listen-client-urls` and `advertise-client-urls` now.

**to reviewers:**
-  previously DM-master listening `8261` for clients request, now another port is needed for peers request. I use `8291` as the default peer port, do you think is it OK?
- this PR not support join a member to a existing cluster, and I'll add it in late PR
- this PR only embed etcd into DM-master, no leader election exists yet, and I'll add it in late PR
- this PR not handle etcd's log, and I'll handle it after ectd in `tidb` and `tidb-tool` upgraded

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported variable/fields change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `dm/dm-ansible`
 - Need to be included in the release note

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@WangXiangUSTC @amyangfei PTAL

File: dm/master/config.go [hunk 1]
```diff
@@ -228,3 +288,59 @@ func (c *Config) Reload() error {
...
+// genEmbedEtcdConfig generates the configuration needed by embed etcd.
+func (c *Config) genEmbedEtcdConfig() (*embed.Config, error) {
+	cfg := embed.NewConfig()
+	cfg.Name = c.Name
+	cfg.Dir = c.DataDir
+
+	// reuse the previous master-addr as the client listening URL.
+	cURL, err := parseURLs(c.MasterAddr)
+	if err != nil {
+		return nil, terror.ErrMasterGenEmbedEtcdConfigFail.Delegate(err, "--listen-client-urls/--advertise-client-urls")
```
Comment (WangXiangUSTC) [inline dm/master/config.go]: we don't have `--listen-client-urls` or `--advertise-client-urls` in dm's config, the user may don't know how to solve this error, how about making it clear to let the user know the `MasterAddr` is wrong

File: dm/master/config.go [hunk 2]
```diff
@@ -228,3 +288,59 @@ func (c *Config) Reload() error {
...
+	}
+	cfg.LCUrls = cURL
+	cfg.ACUrls = cURL
+
+	cfg.LPUrls, err = parseURLs(c.PeerUrls)
+	if err != nil {
+		return nil, terror.ErrMasterGenEmbedEtcdConfigFail.Delegate(err, "--listen-peer-urls")
```
Comment (WangXiangUSTC) [inline dm/master/config.go]: ditto

File: dm/master/config.go [hunk 3]
```diff
@@ -228,3 +288,59 @@ func (c *Config) Reload() error {
...
+	}
+
+	cfg.APUrls, err = parseURLs(c.AdvertisePeerUrls)
+	if err != nil {
+		return nil, terror.ErrMasterGenEmbedEtcdConfigFail.Delegate(err, "--initial-advertise-peer-urls")
```
Comment (WangXiangUSTC) [inline dm/master/config.go]: ditto

File: dm/master/config.go [hunk 4]
```diff
@@ -228,3 +288,59 @@ func (c *Config) Reload() error {
...
+	if s == "" {
+		return nil, nil
+	}
+
+	items := strings.Split(s, ",")
+	urls := make([]url.URL, 0, len(items))
+	for _, item := range items {
+		u, err := url.Parse(item)
+		if err != nil && (strings.Contains(err.Error(), "missing protocol scheme") ||
+			strings.Contains(err.Error(), "first path segment in URL cannot contain colon")) {
```
Comment (WangXiangUSTC) [inline dm/master/config.go]: how about adding a comment to show two examples, in which two cases will generate these error?

**Comment by WangXiangUSTC** [comment]
```
previously DM-master listening 8261 for clients request, now another port is needed for peers request. I use 8269 as the default peer port, do you think is it OK?
```
some users deploy dm-master and more than one dm-worker in one server, perhaps they use port like 8261, 8262, 8263 ...
how about use another port?

File: dm/master/config.go [hunk 5]
```diff
@@ -202,7 +232,37 @@ func (c *Config) adjust() error {
 		c.RPCRateBurst = DefaultBurst
 	}
 
-	return nil
+	if c.Name == "" {
+		var hostname string
+		hostname, err = os.Hostname()
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply dm/master/config.go]: if using `:=`, this inner `err` will shadow the outer one (declared in L197)

File: dm/master/config.go (hunk 1 above)
Comment (csuzhangxc) [reply dm/master/config.go]: updated in https://github.com/pingcap/dm/pull/344/commits/31436e8155c1fac9c59e8a2bc5d78c8399d651a6.

File: dm/master/config.go (hunk 2 above)
Comment (csuzhangxc) [reply dm/master/config.go]: updated in https://github.com/pingcap/dm/pull/344/commits/31436e8155c1fac9c59e8a2bc5d78c8399d651a6

File: dm/master/config.go (hunk 3 above)
Comment (csuzhangxc) [reply dm/master/config.go]: updated in https://github.com/pingcap/dm/pull/344/commits/31436e8155c1fac9c59e8a2bc5d78c8399d651a6

File: dm/master/config.go (hunk 4 above)
Comment (csuzhangxc) [reply dm/master/config.go]: add some comments in https://github.com/pingcap/dm/pull/344/commits/31436e8155c1fac9c59e8a2bc5d78c8399d651a6.;

**Comment by WangXiangUSTC** [comment]
LGTM

**Comment by csuzhangxc** [comment]
@amyangfei PTAL.  we are adding another integration test case `Integration Test import_gorout…`, and it should be failed in this PR, that's not a problem.

File: dm/master/etcd.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	if httpHandles != nil {
+		cfg.UserHandlers = httpHandles
+	}
+
+	e, err := embed.StartEtcd(cfg)
+	if err != nil {
+		return nil, terror.ErrMasterStartEmbedEtcdFail.Delegate(err)
+	}
+
+	timeout := time.Minute
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply dm/master/etcd.go]: changed in https://github.com/pingcap/dm/pull/344/commits/a67bcf93a1d760a26269c63b579af4243473784c.

File: dm/master/server.go [hunk 7]
```diff
@@ -95,43 +94,55 @@ func NewServer(cfg *Config) *Server {
...
 
 // Start starts to serving
-func (s *Server) Start() error {
-	var err error
+func (s *Server) Start(ctx context.Context) (err error) {
+	// create clients to DM-workers
+	for _, workerAddr := range s.cfg.DeployMap {
+		s.workerClients[workerAddr], err = workerrpc.NewGRPCClient(workerAddr)
+		if err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply dm/master/server.go]: done in https://github.com/pingcap/dm/pull/344/commits/a67bcf93a1d760a26269c63b579af4243473784c.

**Pull request merged**